Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeActionHandler.cpp

```diff
@@ -0,0 +1,132 @@
+            EditorEntityContextRequestBus::BroadcastResult(editorEntityContextId, &EditorEntityContextRequests::GetEditorEntityContextId);
+
+            ActionManagerRegistrationNotificationBus::Handler::BusConnect();
+            ComponentModeFramework::EditorComponentModeNotificationBus::Handler::BusConnect(editorEntityContextId);
+        }
+    }
+
+    ComponentModeActionHandler::~ComponentModeActionHandler()
+    {
+        if (IsNewActionManagerEnabled())
+        {
+            ComponentModeFramework::EditorComponentModeNotificationBus::Handler::BusDisconnect();
+            ActionManagerRegistrationNotificationBus::Handler::BusDisconnect();
+        }
+    }
+
+    void ComponentModeActionHandler::OnActionContextModeRegistrationHook()
+    {
+        m_actionManagerInterface = AZ::Interface<AzToolsFramework::ActionManagerInterface>::Get();
+        AZ_Assert(
+            m_actionManagerInterface,
+            "ComponentModeActionHandler - could not get ActionManagerInterface on ComponentModeActionHandler "
+            "OnActionUpdaterRegistrationHook.");
+
+        m_actionManagerInterface->RegisterActionContextMode(EditorMainWindowActionContextIdentifier, "AnyComponentMode");
+    }
+
+    void ComponentModeActionHandler::OnActionUpdaterRegistrationHook()
+    {
+        m_actionManagerInterface->RegisterActionUpdater(ComponentModeChangedUpdaterIdentifier);
+    }
+
+    void ComponentModeActionHandler::OnActionRegistrationHook()
+    {
+        m_hotKeyManagerInterface = AZ::Interface<AzToolsFramework::HotKeyManagerInterface>::Get();
+        AZ_Assert(
+            m_hotKeyManagerInterface,
+            "ComponentModeActionHandler - could not get HotKeyManagerInterface on ComponentModeActionHandler OnActionRegistrationHook.");
+
+        // Add default actions for every Component Mode
+        
+        // Exit Component Mode
+        {
+            constexpr AZStd::string_view actionIdentifier = "o3de.action.componentMode.end";
+            AzToolsFramework::ActionProperties actionProperties;
+            actionProperties.m_name = "Done";
+            actionProperties.m_description = "Return to normal viewport editing";
+            actionProperties.m_category = "Component Modes";
+
+            m_actionManagerInterface->RegisterAction(
+                EditorMainWindowActionContextIdentifier,
+                actionIdentifier,
+                actionProperties,
+                []
+                {
+                    ComponentModeFramework::ComponentModeSystemRequestBus::Broadcast(
+                        &ComponentModeFramework::ComponentModeSystemRequests::EndComponentMode);
+                }
+            );
+
+            // Only add this to the component modes, not default.
+            m_actionManagerInterface->AssignModeToAction("SomeMode", actionIdentifier);
+
+            m_hotKeyManagerInterface->SetActionHotKey(actionIdentifier, "Esc");
```

Comment:
question: One thing I'd like to know is does this still allow actions to 'stack' - so in the example of the Spline Component, if you enter Component Mode and have no vertices selected, and press Esc, you leave Component Mode. However if you have a vertex selected, and press Esc, if will first deselect the Vertex, and then pressing Esc again will leave Component Mode. This might not be in this change but having the ability for shortcuts to dynamically change based on the current state of the mode is super important and I just want to double check we can still support that with the new system. Thanks!

Reply:
Okay thanks Danilo, using the Spline Component as an example of how we want things to work would be great (we have similar behavior in White Box too) where Esc acts as 'back' wherever you are in the current 'stack' of operations and it definitely matches with what uses expect should happen. Having 'Esc' always leave Component Mode actually feels horrible even if in theory it seems to make sense.